lib/http_server.ex

```diff
@@ -15,19 +18,87 @@ defmodule HTTPServer do
   end
 
   defp serve(socket) do
+    request =
+      socket
+      |> read_line()
+      |> parse()
+
+    response =
+      request
+      |> build_response()
+      |> format_response()
+
     socket
-    |> read_line()
-    |> write_line(socket)
+    |> write_line(response)
     |> serve()
   end
 
   defp read_line(socket) do
-    {:ok, data} = :gen_tcp.recv(socket, 0)
-    data
+    case :gen_tcp.recv(socket, 0) do
+      {:ok, data} ->
+        data
+
+      {:error, :econnrefused} ->
+        exit(:normal)
+
+      {:error, :closed} ->
+        exit(:normal)
+    end
   end
 
-  defp write_line(data, socket) do
+  defp write_line(socket, data) do
     :gen_tcp.send(socket, data)
     socket
   end
+
+  def parse(message) do
```

Comment:
```suggestion
  def parse_request(message) do
```
if I'm reading it correctly, this is parsing the request - is that right? if so, this might help with clarity so I know at first glance.

Reply:
also, if it's parsing the request, maybe parameter named `request` is more explicit?